tracker/derived.go

```diff
@@ -0,0 +1,235 @@
+package tracker
+
+import (
+	"github.com/vsariola/sointu"
+	"iter"
+	"slices"
+)
+
+/*
+	from modelData we can derive useful information that can be cached for performance
+	or easy access, because of nested iterations over the score or patch data.
+	i.e. this needs to update when the model changes, and only then.
+*/
+
+type (
+	derivedForUnit struct {
+		unit       *sointu.Unit
+		instrument *sointu.Instrument
+		sends      []*sointu.Unit
+	}
+
+	derivedForTrack struct {
+		instrumentRange          []int
+		tracksWithSameInstrument []int
+		title                    string
+	}
+
+	derivedModelData struct {
+		// map unit by ID
+		forUnit map[int]derivedForUnit
+		// map track by index
+		forTrack map[int]derivedForTrack
+	}
+)
+
+// public access functions
+
+func (m *Model) forUnitById(id int) *derivedForUnit {
+	forUnit, ok := m.derived.forUnit[id]
+	if !ok {
+		return nil
+	}
+	return &forUnit
+}
+
+func (m *Model) InstrumentForUnit(id int) *sointu.Instrument {
+	fu := m.forUnitById(id)
+	if fu == nil {
+		return nil
+	}
+	return fu.instrument
+}
+
+func (m *Model) UnitById(id int) *sointu.Unit {
+	fu := m.forUnitById(id)
+	if fu == nil {
+		return nil
+	}
+	return fu.unit
+}
+
+func (m *Model) SendTargetsForUnit(id int) []*sointu.Unit {
+	fu := m.forUnitById(id)
+	if fu == nil {
+		return nil
+	}
+	return fu.sends
+}
+
+func (m *Model) forTrackByIndex(index int) *derivedForTrack {
+	forTrack, ok := m.derived.forTrack[index]
+	if !ok {
+		return nil
+	}
+	return &forTrack
+}
+
+func (m *Model) TrackTitle(index int) string {
+	ft := m.forTrackByIndex(index)
+	if ft == nil {
+		return ""
+	}
+	return ft.title
+}
+
+// public getters with further model information
+
+func (m *Model) TracksWithSameInstrumentAsCurrent() []int {
+	currentTrack := m.d.Cursor.Track
+	return m.derived.forTrack[currentTrack].tracksWithSameInstrument
```

Comment:
Forget what I said, this won't crash because it's a map of value types. So, this is fine. It'll just return nil slice if it cant find the key in the map.

Reply:
ah. is there actually any difference between a map[int] and a normal slice, if all possible int keys are 0..<max value> anyway? what should we prefer?

Reply:
Array loads and stores are a bit more performant, but this is insignificant. Maps are a bit more robust in the sense that if you try fetch a key that does not exist in the map, it returns the default value. Arrays can cause panics if the index is out of bounds. But checking that index is within bounds is also easy.

Tl;DR probably doesn't matter; given all indices are always linear range 0..N, arrays a bit more natural choice.

Reply:
forTrack / forPattern now use arrays